### PR TITLE
Parametrize the quadrature rule, add Gauss-Lobatto rules

### PR DESCRIPTION
**Pull request by filiatra**

### CODE REVIEW

File: src/gsAssembler/gsQuadRule.h [hunk 1]
```diff
@@ -69,9 +57,21 @@ class gsQuadRule
      *
      */
     virtual void setNodes( gsVector<index_t> const & numNodes,
-                           unsigned digits = std::numeric_limits<T>::digits10 )
+                           unsigned digits = REAL_DIG )
```
Comment (stefantakacs) [inline src/gsAssembler/gsQuadRule.h]: we might prfer c++ over c

File: src/gsAssembler/gsQuadrature.h [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+    /// Constructs a quadrature rule based on input \a options
+    template<class T>
+    static inline gsQuadRule<T> get(int qu, gsVector<index_t> const & numNodes)
+    {
+        switch (qu)
+        {
+        case GaussLobatto :
+            return gsLobattoRule<T>(numNodes);
+        default://GaussLegendre
```
Comment (stefantakacs) [inline src/gsAssembler/gsQuadrature.h]: invalid values shoud invoke GISMO_ERROR

File: src/gsAssembler/gsQuadrature.h [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        const int qu = options.askInt("quRule", GaussLegendre);
+        const T       quA = options.getReal("quA");
+        const index_t quB = options.getInt ("quB");
+        const gsVector<index_t> nnodes = numNodes(basis,quA,quB,fixDir);
+        return get<T>(qu, nnodes);
+    }
+
+    /// Constructs a quadrature rule based on input \a options
+    template<class T>
+    static inline gsQuadRule<T> get(int qu, gsVector<index_t> const & numNodes)
```
Comment (stefantakacs) [inline src/gsAssembler/gsQuadrature.h]: why "int"? shouln't it be "rule"?

File: src/gsAssembler/gsQuadrature.h [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            return gsGaussRule<T>(numNodes);
+        };
+    }
+
+    /// Computes and integer quA*deg_i + quB where deg_i is the degree
+    /// of \a basis
+    template<class T>
+    static gsVector<index_t> numNodes(const gsBasis<T> & basis,
+                               const T quA, const int quB, int fixDir = -1)
+    {
```
Comment (stefantakacs) [inline src/gsAssembler/gsQuadrature.h]: why not index_t?

many followups...

File: src/gsAssembler/gsQuadRule.h (hunk 1 above)
Comment (filiatra) [reply src/gsAssembler/gsQuadRule.h]: This is defined in gsMath.h and is equal to std::numeric_limits<T>::digits10

File: src/gsAssembler/gsQuadrature.h (hunk 3 above)
Comment (filiatra) [reply src/gsAssembler/gsQuadrature.h]: GISMO_ERROR will be invoked for invalid values, so passing int 1 or 2 is perfectly fine, and I want ot allow it, as it is the same as the enum

File: src/gsAssembler/gsQuadrature.h (hunk 4 above)
Comment (filiatra) [reply src/gsAssembler/gsQuadrature.h]: I want to make sure that -1 is a valid value (currently index_t=int but we can imagine index_t=unsigned)

File: src/gsAssembler/gsQuadrature.h (hunk 3 above)
Comment (stefantakacs) [reply src/gsAssembler/gsQuadrature.h]: This seems to be a general question. Doing the same, e.g., for InterfaceStrategy and BoundaryStrategy in assemblers and dof mapper etc. would also save many casts...

Comment (filiatra) [reply src/gsAssembler/gsQuadrature.h]: yes, in fact in the assemblers these will go away completely, they are in the options instead..

**Pull request merged**